cdc/sink/dmlsink/mq/mq_dml_sink.go

```diff
@@ -119,35 +120,49 @@ func newDMLSink(
 
 // WriteEvents writes events to the sink.
 // This is an asynchronously and thread-safe method.
-func (s *dmlSink) WriteEvents(rows ...*dmlsink.RowChangeCallbackableEvent) error {
+func (s *dmlSink) WriteEvents(txns ...*dmlsink.CallbackableEvent[*model.SingleTableTxn]) error {
 	s.alive.RLock()
 	defer s.alive.RUnlock()
 	if s.alive.isDead {
 		return errors.Trace(errors.New("dead dmlSink"))
 	}
-
-	for _, row := range rows {
-		if row.GetTableSinkState() != state.TableSinkSinking {
+	// merge the split row callback into one callback
+	mergedCallback := func(outCallback func(), totalCount uint64) func() {
+		var acked atomic.Uint64
+		return func() {
+			if acked.Add(1) == totalCount {
+				outCallback()
+			}
+		}
+	}
+	for _, txn := range txns {
+		if txn.GetTableSinkState() != state.TableSinkSinking {
 			// The table where the event comes from is in stopping, so it's safe
 			// to drop the event directly.
-			row.Callback()
+			txn.Callback()
 			continue
 		}
-		topic := s.alive.eventRouter.GetTopicForRowChange(row.Event)
-		partitionNum, err := s.alive.topicManager.GetPartitionNum(s.ctx, topic)
-		if err != nil {
-			return errors.Trace(err)
-		}
-		partition := s.alive.eventRouter.GetPartitionForRowChange(row.Event, partitionNum)
-		// This never be blocked because this is an unbounded channel.
-		s.alive.worker.msgChan.In() <- mqEvent{
-			key: TopicPartitionKey{
-				Topic: topic, Partition: partition,
-			},
-			rowEvent: row,
+		callback := mergedCallback(txn.Callback, uint64(len(txn.Event.Rows)))
+		for _, row := range txn.Event.Rows {
+			topic := s.alive.eventRouter.GetTopicForRowChange(row)
```

Comment:
It looks that `for one transactions, all rows belongs to the same table, have the same commitTs, so should be delivered to the same topic and partition.`

Reply:
the topic and partition are based on row value and dispatch rule?